cilium-cli/connectivity/tests/errors.go

```
@@ -157,10 +157,18 @@ func (n *noErrorsInLogs) Run(ctx context.Context, t *check.Test) {
 				// Do not check for container restarts for Cilium v1.16 and earlier.
 				ignore := n.ciliumVersion.LT(semver.MustParse("1.17.0"))
 
-				// Ignore Cilium operator restarts for the moment, as they can
-				// legitimately happen in case it loses the leader election due
-				// to temporary control plane blips.
-				ignore = ignore || container == "cilium-operator"
+				// Check for legitimate cilium-operator restart patterns
+				if container == "cilium-operator" && restarts > 0 {
+					var logs bytes.Buffer
+					err := client.GetLogs(ctx, pod.Namespace, pod.Name, container, prevOpts, &logs)
+					if err != nil {
+						a.Logf("Warning: Could not retrieve previous logs to check for legitimate operator restart patterns: %s", err)
+					} else {
+						if hasLegitimateOperatorRestartPatterns(logs.Bytes()) {
+							ignore = true
+						}
```

Does this mean that if there is a serious error and an acceptable error, the serious error will be ignored just because "Leader election lost" appears somewhere in the logs?

---

@joestringer Good catch! Leader election lost will suppress only the restart-count failure. The error logs (previous and current) are always to be checked otherwise we could accidentally ignore other (possibly serious) errors.

---

Still not addressed.

---

The logic sets `ignore = true` only for the restart-count, then immediately checks previous logs for errors, and separately always checks current logs:

- Previous logs check (restart-count suppression + error scan):
```160:170:cilium-cli/connectivity/tests/errors.go
// Check for legitimate cilium-operator restart patterns
if container == "cilium-operator" && restarts > 0 {
    var prevLogs bytes.Buffer
    err := client.GetLogs(ctx, pod.Namespace, pod.Name, container, prevOpts, &prevLogs)
    if err != nil {
        a.Logf("Warning: Could not retrieve previous logs to check for legitimate operator restart patterns: %s", err)
    } else if hasLegitimateOperatorRestartPatterns(prevLogs.Bytes()) {
        // Ignore the restart count, but still verify there are no error logs in previous logs
        ignore = true
        n.checkErrorsInLogs(id, prevLogs.Bytes(), a, &prevOpts)
    }
}
```

- Current logs check (always scanned for errors):
```181:187:cilium-cli/connectivity/tests/errors.go
var logs bytes.Buffer
err := client.GetLogs(ctx, pod.Namespace, pod.Name, container, opts, &logs)
if err != nil {
    a.Fatalf("Error reading Cilium logs: %s", err)
}
n.checkErrorsInLogs(id, logs.Bytes(), a, &opts)
```

Thanks for the follow-up @joestringer! Could you clarify what change you’d like to see beyond this?
- Should we avoid suppressing the restart-count entirely (warn instead)?
- Or is there another behavior you want adjusted?

---

I took a fresh look and proposed a path forward below.